From b43e2d7b11646d047a2d504bf3790900589aa7e3 Mon Sep 17 00:00:00 2001
From: zeetim <support@zeetim.com>
Date: Tue, 11 Mar 2025 14:48:23 +0100
Subject: [PATCH] Makefile: Add MOK_POLICY variable to build options to change
 default policy value.

Signed-off-by: zeetim <support@zeetim.com>
---
 Makefile  | 5 +++++
 globals.c | 4 ++++
 2 files changed, 9 insertions(+)

diff --git a/Makefile b/Makefile
index daf8114..03941b6 100644
--- a/Makefile
+++ b/Makefile
@@ -80,6 +80,11 @@ ifneq ($(origin FALLBACK_VERBOSE_WAIT), undefined)
 	CFLAGS += -DFALLBACK_VERBOSE_WAIT=$(FALLBACK_VERBOSE_WAIT)
 endif
 
+ifneq ($(origin MOK_POLICY), undefined)
+$(warning Setting default MOK policy to $(MOK_POLICY))
+	CFLAGS += -DMOK_POLICY_DEFAULT=$(MOK_POLICY)
+endif
+
 all: confcheck certcheck $(TARGETS)
 
 confcheck:
diff --git a/globals.c b/globals.c
index 1119712..436bcce 100644
--- a/globals.c
+++ b/globals.c
@@ -31,7 +31,11 @@ UINT8 user_insecure_mode;
 UINTN hsi_status = 0;
 UINT8 ignore_db;
 UINT8 trust_mok_list;
+#ifdef MOK_POLICY_DEFAULT
+UINT8 mok_policy = MOK_POLICY_DEFAULT;
+#else
 UINT8 mok_policy = 0;
+#endif
 
 UINT32 verbose = 0;
 
-- 
2.39.5

