From a776436a51aafffbf945f727c56cbba5798cb05b Mon Sep 17 00:00:00 2001
From: Guillaume Chaye <guillaume.chaye@zeetim.com>
Date: Wed, 20 Aug 2025 11:39:03 +0200
Subject: [PATCH] Makefile: Add MOK_POLICY variable to build options to change
 default policy value.

Signed-off-by: Guillaume Chaye <guillaume.chaye@zeetim.com>
---
 Makefile  | 5 +++++
 globals.c | 4 ++++
 2 files changed, 9 insertions(+)

diff --git a/Makefile b/Makefile
index 9a39cd1..3d2272d 100644
--- a/Makefile
+++ b/Makefile
@@ -69,6 +69,11 @@ ifneq ($(origin FALLBACK_VERBOSE_WAIT), undefined)
 	CFLAGS += -DFALLBACK_VERBOSE_WAIT=$(FALLBACK_VERBOSE_WAIT)
 endif
 
+ifneq ($(origin MOK_POLICY), undefined)
+$(warning Setting default MOK policy to $(MOK_POLICY))
+	CFLAGS += -DMOK_POLICY_DEFAULT=$(MOK_POLICY)
+endif
+
 all: confcheck certcheck $(TARGETS)
 
 confcheck:
diff --git a/globals.c b/globals.c
index 1119712..436bcce 100644
--- a/globals.c
+++ b/globals.c
@@ -31,7 +31,11 @@ UINT8 user_insecure_mode;
 UINTN hsi_status = 0;
 UINT8 ignore_db;
 UINT8 trust_mok_list;
+#ifdef MOK_POLICY_DEFAULT
+UINT8 mok_policy = MOK_POLICY_DEFAULT;
+#else
 UINT8 mok_policy = 0;
+#endif
 
 UINT32 verbose = 0;
 
-- 
2.47.2

